dev/fattest.simplicity/src/componenttest/rules/repeater/FeatureReplacementAction.java

```diff
@@ -401,9 +432,126 @@ public boolean isEnabled() {
+            }
+        }
+        for (String alwaysAddFeature : alwaysAddFeatures) {
+            if (alwaysAddFeature.toLowerCase().startsWith(baseFeature.toLowerCase())) {
+                replaceFeature = alwaysAddFeature;
+            }
+        }
+        if (replaceFeature != null) {
+            Log.info(c, methodName, "Replace feature [ " + originalFeature + " ] with [ " + replaceFeature + " ]");
+            return replaceFeature;
+        }
+        // We need to check that the feature passed is an EE7/EE8 feature which could have a name change on EE9 that doesn't match
+        // the original feature name it replaces. We also check vice versa if the feature is an EE9 feature that involves a name change
+        // to update from EE9 to EE7/EE8
+        Log.info(c, methodName, "No feature replacement found for [ " + originalFeature + " ]. Verifying if feature name was changed on EE9.");
+        // Reset base feature to not include the "-"
+        baseFeature = baseFeature.substring(0, dashOffset);
+        String featureReplacement = null;
+        for (Map.Entry<String, String> nameChangeEntry : featuresWithNameChangeOnEE9.entrySet()) {
+            if (nameChangeEntry.getValue().equalsIgnoreCase(baseFeature)) {
+                featureReplacement = nameChangeEntry.getKey();
+                Log.info(c, methodName, "Replace EE9 feature [ " + baseFeature + " ] with feature [ " + featureReplacement + " ]");
+                baseFeature = featureReplacement;
+                break;
+            }
+            if (nameChangeEntry.getKey().equalsIgnoreCase(baseFeature)) {
+                featureReplacement = nameChangeEntry.getValue();
+                Log.info(c, methodName, "Replace base feature [ " + baseFeature + " ] with EE9 feature [ " + featureReplacement + " ]");
+                baseFeature = featureReplacement;
+                break;
+            }
+        }
+        if (featureReplacement == null) {
+            Log.info(c, methodName, "Remove feature [ " + originalFeature + " ]: No replacement is available");
+            return null;
+        }
+        baseFeature += "-";
+        // Re-check the features with the name changes
+        for (String replacementFeature : replacementFeatures) {
+            if (replacementFeature.toLowerCase().contains(baseFeature.toLowerCase())) {
+                Log.info(c, methodName, "Replace feature [ " + originalFeature + " ] with [ " + replacementFeature + " ]");
+                return replacementFeature;
+            }
+        }
+        // We may need to remove a feature without adding any replacement
+        // (e.g. jsonb-1.0 is EE8 only) so in this case return null
+        Log.info(c, methodName, "Remove feature [ " + originalFeature + " ]: No replacement is available");
+        return null;
+    }
+
+    /**
+     * Goes through all the servers and clients making the specified feature changes
+     */
+    private void setFeatures() throws Exception {
```

Comment:
If you moved setFeatures to right before cleanup() I think the diff would be much cleaner.

Reply:
Will try that out!